build fails with Java 12

Running _gradlew clean build -PallTests_ with Java 12.0.1 on Windows 10, I get test failures : 
- in com.tngtech.archunit.core.importer.ClassFileImporterTest , [imports_simple_enum](https://github.com/TNG/ArchUnit/blob/783e72b5409ae219e3540ec15d993778a882c0f8/archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterTest.java#L274) :

		java.lang.AssertionError: [classes] 
		Expected size:<2> but was:<3> in:
		<[JavaClass{name='java.io.Serializable'},
			JavaClass{name='java.lang.Comparable'},
			JavaClass{name='java.lang.constant.Constable'}]>
			at com.tngtech.archunit.testutil.Assertions$JavaClassesAssertion.matchExactly(Assertions.java:282)
			at com.tngtech.archunit.testutil.Assertions$JavaClassesAssertion.matchInAnyOrder(Assertions.java:274)
			at com.tngtech.archunit.testutil.Assertions$JavaClassesAssertion.matchInAnyOrder(Assertions.java:278)
			at com.tngtech.archunit.core.importer.ClassFileImporterTest.imports_simple_enum(ClassFileImporterTest.java:283)
			at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)


--> Related to the new Java 12 [Constable interface](https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/lang/constant/Constable.html)

- in com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest, [resideInAnyPackage](https://github.com/TNG/ArchUnit/blob/783e72b5409ae219e3540ec15d993778a882c0f8/archunit/src/test/java/com/tngtech/archunit/lang/syntax/elements/GivenMembersDeclaredInClassesThatTest.java#L175) : 

		java.lang.AssertionError: [Actual JavaMembers] 
		Actual and expected should have same size but actual size is:
		 <131>
		while expected is:
		 <120>
		Actual was:
		<[JavaConstructor{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.<init>()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.and_conjunction()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.annotatedWith_Retention_Source_is_rejected()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.areAnnotatedWith_predicate()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.areAnnotatedWith_type()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.areAnnotatedWith_typeName()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.areAssignableFrom_predicate()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.areAssignableFrom_type()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.areAssignableFrom_typeName()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.areAssignableTo_predicate()},
			JavaMethod{com.tngtech.archunit.lang.syntax.elements.GivenMembersDeclaredInClassesThatTest.areAssignableTo_type()},

- in com.tngtech.archunit.integration.ExamplesIntegrationTest, [FrozenRulesTest](https://github.com/TNG/ArchUnit/blob/783e72b5409ae219e3540ec15d993778a882c0f8/archunit-integration-test/src/test/java/com/tngtech/archunit/integration/ExamplesIntegrationTest.java#L565)

		org.opentest4j.AssertionFailedError: Test class com.tngtech.archunit.exampletest.FrozenRulesTest hasn't produced the correct violations:

		Test has not failed as expected: 
		Result from comparing violations: 
		Unexpected message lines: [Class <com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService> implements interface <com.tngtech.archunit.example.service.ServiceInterface> in (DaoCallingService.java:0), Class <com.tngtech.archunit.example.service.impl.ServiceImplementation> implements interface <com.tngtech.archunit.example.service.ServiceInterface> in (ServiceImplementation.java:0), Constructor <com.tngtech.archunit.example.SomeMediator.<init>(com.tngtech.archunit.example.service.ServiceViolatingLayerRules)> has parameter of type <com.tngtech.archunit.example.service.ServiceViolatingLayerRules> in (SomeMediator.java:0), Field <com.tngtech.archunit.example.SomeMediator.service> has type <com.tngtech.archunit.example.service.ServiceViolatingLayerRules> in (SomeMediator.java:0), Field <com.tngtech.archunit.example.controller.SomeController.otherService> has type <com.tngtech.archunit.example.service.ServiceViolatingLayerRules> in (SomeController.java:0), Field <com.tngtech.archunit.example.controller.SomeController.service> has type <com.tngtech.archunit.example.service.ServiceViolatingDaoRules> in (SomeController.java:0), Field <com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService.service> has type <com.tngtech.archunit.example.service.ServiceViolatingLayerRules> in (DaoCallingService.java:0), Field <com.tngtech.archunit.example.service.ServiceViolatingDaoRules.myEntityManager> has type <com.tngtech.archunit.example.service.ServiceViolatingDaoRules$MyEntityManager> in (ServiceViolatingDaoRules.java:0), Method <com.tngtech.archunit.example.SomeMediator.violateLayerRulesIndirectly()> calls method <com.tngtech.archunit.example.service.ServiceViolatingLayerRules.doSomething()> in (SomeMediator.java:15), Method <com.tngtech.archunit.example.controller.SomeGuiController.callServiceLayer()> calls constructor <com.tngtech.archunit.example.service.ServiceHelper.<init>()> in (SomeGuiController.java:7), Method <com.tngtech.archunit.example.controller.SomeGuiController.callServiceLayer()> calls constructor <com.tngtech.archunit.example.service.ServiceHelper.<init>(java.lang.String)> in (SomeGuiController.java:8), Method <com.tngtech.archunit.example.controller.SomeGuiController.callServiceLayer()> gets field <com.tngtech.archunit.example.service.ServiceHelper.insecure> in (SomeGuiController.java:10), Method <com.tngtech.archunit.example.controller.SomeGuiController.callServiceLayer()> gets field <com.tngtech.archunit.example.service.ServiceHelper.properlySecured> in (SomeGuiController.java:11), Method <com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService.violateLayerRules()> calls method <com.tngtech.archunit.example.service.ServiceViolatingLayerRules.doSomething()> in (DaoCallingService.java:14)]
		Result from handling violations: 
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.SomeMediator'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingLayerRules'}, lineNumber=0, description=Constructor <com.tngtech.archunit.example.SomeMediator.<init>(com.tngtech.archunit.example.service.ServiceViolatingLayerRules)> has parameter of type <com.tngtech.archunit.example.service.ServiceViolatingLayerRules> in (SomeMediator.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.SomeMediator'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingLayerRules'}, lineNumber=0, description=Field <com.tngtech.archunit.example.SomeMediator.service> has type <com.tngtech.archunit.example.service.ServiceViolatingLayerRules> in (SomeMediator.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.SomeMediator'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingLayerRules'}, lineNumber=15, description=Method <com.tngtech.archunit.example.SomeMediator.violateLayerRulesIndirectly()> calls method <com.tngtech.archunit.example.service.ServiceViolatingLayerRules.doSomething()> in (SomeMediator.java:15)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.controller.SomeController'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingDaoRules'}, lineNumber=0, description=Field <com.tngtech.archunit.example.controller.SomeController.service> has type <com.tngtech.archunit.example.service.ServiceViolatingDaoRules> in (SomeController.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.controller.SomeController'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingLayerRules'}, lineNumber=0, description=Field <com.tngtech.archunit.example.controller.SomeController.otherService> has type <com.tngtech.archunit.example.service.ServiceViolatingLayerRules> in (SomeController.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.controller.SomeGuiController'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceHelper'}, lineNumber=10, description=Method <com.tngtech.archunit.example.controller.SomeGuiController.callServiceLayer()> gets field <com.tngtech.archunit.example.service.ServiceHelper.insecure> in (SomeGuiController.java:10)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.controller.SomeGuiController'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceHelper'}, lineNumber=11, description=Method <com.tngtech.archunit.example.controller.SomeGuiController.callServiceLayer()> gets field <com.tngtech.archunit.example.service.ServiceHelper.properlySecured> in (SomeGuiController.java:11)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.controller.SomeGuiController'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceHelper'}, lineNumber=7, description=Method <com.tngtech.archunit.example.controller.SomeGuiController.callServiceLayer()> calls constructor <com.tngtech.archunit.example.service.ServiceHelper.<init>()> in (SomeGuiController.java:7)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.controller.SomeGuiController'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceHelper'}, lineNumber=8, description=Method <com.tngtech.archunit.example.controller.SomeGuiController.callServiceLayer()> calls constructor <com.tngtech.archunit.example.service.ServiceHelper.<init>(java.lang.String)> in (SomeGuiController.java:8)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceInterface'}, lineNumber=0, description=Class <com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService> implements interface <com.tngtech.archunit.example.service.ServiceInterface> in (DaoCallingService.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingLayerRules'}, lineNumber=0, description=Field <com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService.service> has type <com.tngtech.archunit.example.service.ServiceViolatingLayerRules> in (DaoCallingService.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingLayerRules'}, lineNumber=14, description=Method <com.tngtech.archunit.example.persistence.layerviolation.DaoCallingService.violateLayerRules()> calls method <com.tngtech.archunit.example.service.ServiceViolatingLayerRules.doSomething()> in (DaoCallingService.java:14)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingDaoRules'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceViolatingDaoRules$MyEntityManager'}, lineNumber=0, description=Field <com.tngtech.archunit.example.service.ServiceViolatingDaoRules.myEntityManager> has type <com.tngtech.archunit.example.service.ServiceViolatingDaoRules$MyEntityManager> in (ServiceViolatingDaoRules.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.service.impl.ServiceImplementation'}, targetClass=JavaClass{name='com.tngtech.archunit.example.service.ServiceInterface'}, lineNumber=0, description=Class <com.tngtech.archunit.example.service.impl.ServiceImplementation> implements interface <com.tngtech.archunit.example.service.ServiceInterface> in (ServiceImplementation.java:0)}
		Result from comparing violations: 
		Unexpected message lines: [Class <com.tngtech.archunit.example.service.ServiceViolatingDaoRules$MyEntityManager> implements interface <javax.persistence.EntityManager> in (ServiceViolatingDaoRules.java:0), Field <com.tngtech.archunit.example.persistence.first.dao.jpa.SomeJpa.entityManager> has type <javax.persistence.EntityManager> in (SomeJpa.java:0), Field <com.tngtech.archunit.example.persistence.second.dao.jpa.OtherJpa.entityManager> has type <javax.persistence.EntityManager> in (OtherJpa.java:0), Field <com.tngtech.archunit.example.service.ServiceViolatingDaoRules.myEntityManager> has type <com.tngtech.archunit.example.service.ServiceViolatingDaoRules$MyEntityManager> in (ServiceViolatingDaoRules.java:0), Method <com.tngtech.archunit.example.persistence.first.dao.jpa.SomeJpa.findById(long)> calls method <javax.persistence.EntityManager.find(java.lang.Class, java.lang.Object)> in (SomeJpa.java:20), Method <com.tngtech.archunit.example.persistence.second.dao.OtherDao.getEntityManager()> has return type <javax.persistence.EntityManager> in (OtherDao.java:0), Method <com.tngtech.archunit.example.persistence.second.dao.jpa.OtherJpa.findById(long)> calls method <javax.persistence.EntityManager.find(java.lang.Class, java.lang.Object)> in (OtherJpa.java:19), Method <com.tngtech.archunit.example.persistence.second.dao.jpa.OtherJpa.getEntityManager()> has return type <javax.persistence.EntityManager> in (OtherJpa.java:0)]
		Result from handling violations: 
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.persistence.first.dao.jpa.SomeJpa'}, targetClass=JavaClass{name='javax.persistence.EntityManager'}, lineNumber=0, description=Field <com.tngtech.archunit.example.persistence.first.dao.jpa.SomeJpa.entityManager> has type <javax.persistence.EntityManager> in (SomeJpa.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.persistence.first.dao.jpa.SomeJpa'}, targetClass=JavaClass{name='javax.persistence.EntityManager'}, lineNumber=20, description=Method <com.tngtech.archunit.example.persistence.first.dao.jpa.SomeJpa.findById(long)> calls method <javax.persistence.EntityManager.find(java.lang.Class, java.lang.Object)> in (SomeJpa.java:20)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.persistence.second.dao.OtherDao'}, targetClass=JavaClass{name='javax.persistence.EntityManager'}, lineNumber=0, description=Method <com.tngtech.archunit.example.persistence.second.dao.OtherDao.getEntityManager()> has return type <javax.persistence.EntityManager> in (OtherDao.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.persistence.second.dao.jpa.OtherJpa'}, targetClass=JavaClass{name='javax.persistence.EntityManager'}, lineNumber=0, description=Field <com.tngtech.archunit.example.persistence.second.dao.jpa.OtherJpa.entityManager> has type <javax.persistence.EntityManager> in (OtherJpa.java:0)}
		Unexpected access: Dependency{originClass=JavaClass{name='com.tngtech.archunit.example.persistence.second.dao.jpa.OtherJpa'}, targetClass=JavaClass{name='javax.persistence.EntityManager'}, lineNumber=0, description=Method <com.tngtech.archunit.example.persistence.second.dao.jpa.OtherJpa.getEntityManager()> has return type <javax.persistence.EntityManager> in (OtherJpa.java:0)}


Not sure how we want to fix this.. should we make assertions more lenient ? If we assert strictly, then Java 12+ will be mandatory to build, and we probably don't want that

--- a/archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterTest.java
+++ b/archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterTest.java
@@ -280,7 +280,7 @@ public class ClassFileImporterTest {
         assertThat(javaClass.getModifiers()).as("modifiers").containsOnly(JavaModifier.PUBLIC, JavaModifier.FINAL);
         assertThat(javaClass.getSuperClass().get()).as("super class").matches(Enum.class);
         assertThat(javaClass.getInterfaces()).as("interfaces").isEmpty();
-        assertThatClasses(javaClass.getAllInterfaces()).matchInAnyOrder(Serializable.class, Comparable.class);
+        assertThatClasses(javaClass.getAllInterfaces()).matchInAnyOrder(Enum.class.getInterfaces());
         assertThat(javaClass.isInterface()).as("is interface").isFalse();
         assertThat(javaClass.isEnum()).as("is enum").isTrue();
     }

--- a/archunit/src/test/java/com/tngtech/archunit/lang/syntax/elements/GivenMembersDeclaredInClassesThatTest.java
+++ b/archunit/src/test/java/com/tngtech/archunit/lang/syntax/elements/GivenMembersDeclaredInClassesThatTest.java
@@ -3,7 +3,7 @@ package com.tngtech.archunit.lang.syntax.elements;
 import java.io.Serializable;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
-import java.lang.reflect.Constructor;
+import java.lang.ref.Reference;
 import java.text.AttributedString;
 import java.util.AbstractList;
 import java.util.ArrayList;
@@ -173,17 +173,18 @@ public class GivenMembersDeclaredInClassesThatTest {
 
     @Test
     public void resideInAnyPackage() {
-        List<JavaMember> members = filterResultOf(members().that().areDeclaredInClassesThat().resideInAnyPackage("..tngtech..", "java.lang.reflect"))
-                .on(getClass(), String.class, Constructor.class);
+        List<JavaMember> members = filterResultOf(members().that().areDeclaredInClassesThat()
+                .resideInAnyPackage("..tngtech..", "java.lang.ref")
+        ).on(getClass(), String.class, Reference.class);
 
-        assertThatMembers(members).matchInAnyOrderMembersOf(getClass(), Constructor.class);
+        assertThatMembers(members).matchInAnyOrderMembersOf(getClass(), Reference.class);
     }
 
     @Test
     public void resideOutsideOfPackages() {
         List<JavaMember> members = filterResultOf(members().that().areDeclaredInClassesThat()
-                .resideOutsideOfPackages("..tngtech..", "java.lang.reflect")
-        ).on(getClass(), String.class, Constructor.class);
+                .resideOutsideOfPackages("..tngtech..", "java.lang.ref")
+        ).on(getClass(), String.class, Reference.class);
 
         assertThatMembers(members).matchInAnyOrderMembersOf(String.class);
     }

